packages/SwingSet/src/vats/comms/delivery.js

```diff
@@ -409,6 +445,11 @@ export function makeDeliveryKit(state, syscall, transmit, clistKit, stateKit) {
       // be handled properly
       markPromiseAsResolved(lpid, rejected, data);
     }
+    const auxResolutions = collector.getResolutions();
+    if (auxResolutions.length > 0) {
```

Comment:
I think the set of resolutions returned by `collector.getResolutions()` will significantly overlap with the `resolutions` we passed in, since we don't tell the collector the full list ahead of time. We know we need to add the promises which we know to be resolved, but the sending remote didn't know about at that moment. But we don't want to duplicate any promises that they *did* know to be resolved.

The `resolutions` that is passed into `handleResolutions` will contain one or more primaries (i.e. promises that we already knew about, which the sending remote is resolving) plus zero or more auxilliaries (ephemeral promise IDs that were just created, by dint of being mentioned in the resolution data of one of the primaries, or that of some other auxilliary that appears earlier in `resolutions`).

The promises that come out of `collector.getResolutions()` might include one or more of the primaries if there's a cycle. It will include all of the auxilliaries. And it might include some number of additional "locally known to be resolved" promises that weren't included in the original message, but need to be included in any outbound messages we send.

To avoid duplicating anything in the message we send to downstream subscribers, I see two approaches.

* Loop through everything that comes back from `collector.getResolutions()`, in order, and add them to the end of `resolutions` iff the lpid isn't already present.
* Start by marking everything as resolved, then use the `resolutionCollector()` in the `forPromise` mode on every promise in `resolutions`. The output of the collector should be the original promise list plus any locally-known-resolved promises. The list might be in a slightly different order: the locally-known-resolved ones may appear in the middle here, whereas in the other approach they'd appear strictly after the auxilliary promises we were given.
* I guess a third approach (a minor variant on the first) is to filter `collector.getResolutions()` by removing anything already present in `resolutions`, before checking the length and doing the concat.

Also, minor nit, I'd remove the `if` and make the `concat` unconditional.

Reply:
I think your analysis is mistaken: all promises that are referenced by the bundle of resolutions passed in will fall into one of two categories -- they are either promises that were truly unresolved from the sender's perspective at the time of sending or promises that are being exported (and thus initially resolved) and then immediate resolved within the resolution bundle.  Consequently, the only resolved resolutions the collector will collect are ones whose resolved state was unknown to the sender, and thus will always be additions to the resolution bundle rather than duplicates.

Reply:
One thing the conditional `concat` does (well, not the conditional `concat` but the conditional itself) is let us log when something has happened and shut up if not.  Minor.

Reply:
What if the unknown-to-the-sender resolution included a reference to itself?

Reply:
Ok, Chip has convinced me that this code is correct.. I had a broken model of what the state of the world looked like when this code is entered. I'll suggest adding the following comments to help me in the future when I forget all of this discussion/convincing and reread this code and jump to the wrong conclusion again:


```js
  function handleResolutions(resolutions) {
    // All promises listed as *targets* of 'resolutions' are unresolved: they
    // are being decided by someone else (the kernel, or a remote). They are
    // either primary (previously known to us) or auxilliary (created, in an
    // unresolved state, as the resolution data of the batch was translated
    // from whatever kernel/remote sent the batch into our local namespace).
    // The resolution data of 'resolutions' may point to both resolved and
    // unresolved promises, but any cycles (pointers into other targets of
    // 'resolutions') will still point to unresolved promises, because we
    // haven't marked them as resolved quite yet.
    const [[primaryLpid]] = resolutions;
    const { subscribers, kernelIsSubscribed } = getPromiseSubscribers(
      primaryLpid,
    );

    // Run the collector before we change the state of our promise table.

    const collector = resolutionCollector();
    for (const resolution of resolutions) {
      const [_lpid, _rejected, data] = resolution;
      collector.forSlots(data.slots);
    }

    // The collector now knows about every previously-resolved promise cited
    // by the resolution data of 'resolutions'. This will never overlap with
    // the targets of 'resolutions'.

    for (const resolution of resolutions) {
      const [lpid, rejected, data] = resolution;
      // rejected: boolean, data: capdata
      insistCapData(data);
      insistLocalType('promise', lpid);
      insistPromiseIsUnresolved(lpid);
      insistDeciderIsComms(lpid);

      // mark it as resolved in the promise table, so later messages to it will
      // be handled properly
      markPromiseAsResolved(lpid, rejected, data);
    }

    // At this point, both the primary and auxillary promises listed as
    // targets of 'resolutions' are marked as resolved.

    const auxResolutions = collector.getResolutions();
    if (auxResolutions.length > 0) {
      // console.log(`@@@ adding ${auxResolutions.length} aux resolutions`);
      // 'auxResolutions' and 'resolutions' are strictly disjoint
      resolutions = resolutions.concat(auxResolutions);
    }
```